Refactor `report_parse_error` in `crate::build`

LighghtEeloo opened the pull request:
<!--
Thanks for your contribution!

We always run tests on the current stable version of Rust.
To avoid unexpected CI result, consider updating Rust if you're using an earlier version.

-->

As noted in #1013, this PR aims to abstract error handling. The only file modified in this PR is `src/build/mod.rs`, where parse error reporting has been refactored out of the function `parse_and_normalize_grammar` into a new public function, `report_parse_error`.

The changes are straightforward: the new `report_parse_error` function is implemented in continuation-passing style, as it accepts a callback of type `impl FnMut(&FileText, pt::Span, &str, io::Error) -> E`. This design allows the caller to produce any desired error type.

That said, I wonder if error handling could be better organized. While this approach works well for all the use cases I've encountered, I suspect it’s far from optimal. Any feedback or suggestions would be greatly appreciated.

LighghtEeloo commented:
Yes, your observation is correct. `report_parse_error()` is called with a different callback in the LSP code, and it's intended to be exposed in the public API of lalrpop_internals. I'll fix the clippy lints in the following commit.

dburgener commented on lalrpop/src/build/mod.rs:
```diff
@@ -266,53 +265,68 @@ fn lalrpop_files<P: AsRef<Path>>(root_dir: P) -> io::Result<Vec<PathBuf>> {
+pub fn report_parse_error<E>(
+    file_text: &FileText,
+    error: parser::ParseError<'_>,
+    mut reporter: impl FnMut(&FileText, pt::Span, &str, io::Error) -> E,
```
This implementation where the callback takes in an io::Error and then turns around and returns that same error (but now it has the generic type E) feels weird.  It looks like in the LSP implementation it just re-wraps up the io::Error.

io::Error was always a sort of weird return here.  The only reason the block that got factored out returned an io::Error is because that's the error type that goes up the stack.  I think rather than do this whole "pass the io::Error around" dance, it would make more sense to just have the internal implementation use io::ErrorKind::InvalidData unconditionally.  I don't care too much to preserve the one UnexpectedEof case, which is only arguably important in the first place.

LighghtEeloo replied:
Good point. If we don't need to keep the io::Error around, then we may consider removing it entirely.

dburgener commented on lalrpop/src/build/mod.rs:
```diff
@@ -266,53 +265,68 @@ fn lalrpop_files<P: AsRef<Path>>(root_dir: P) -> io::Result<Vec<PathBuf>> {
+    }
+}
+
+pub fn report_parse_error<E>(
```
This function needs a doc comment.  One because the whole callback dance is a little weird, and should be explained, but 2, the ultimate goal is for this to become a public export in lalrpop_internals, and we'd like to have it documented for that.

LighghtEeloo replied:
Indeed. I'll add the docs.

LighghtEeloo commented:
Sorry for the late response due to personal reasons. I've updated the git branch to add documentations and removed `io::Error` from the callback parameter.

dburgener reviewed (approved):
This is so much cleaner without the dance passing the error to the callback!

Thanks for the update.  This LGTM.  I'll give another day for any other maintainer reviews before merging.

EDIT: Just saw CI.  This will need a rustfmt run before merge as well.

Pat-Lafon reviewed (approved):
👍
